storage: TestNodeLivenessStatusMap failed under stress

SHA: https://github.com/cockroachdb/cockroach/commits/f38e4c7a2bb3f0e697b25d21bdd8ea67b163c32f

Parameters:
```
TAGS=
GOFLAGS=-race
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=787834&tab=buildLog

```
=== RUN   TestNodeLivenessStatusMap
W180720 07:16:36.400821 123057 server/status/runtime.go:143  Could not parse build timestamp: parsing time "" as "2006/01/02 15:04:05": cannot parse "" as "2006"
I180720 07:16:36.514231 123057 server/server.go:795  [n?] monitoring forward clock jumps based on server.clock.forward_jump_check_enabled
I180720 07:16:36.527088 123057 server/config.go:552  [n?] 1 storage engine initialized
I180720 07:16:36.527161 123057 server/config.go:555  [n?] RocksDB cache size: 128 MiB
I180720 07:16:36.527198 123057 server/config.go:555  [n?] store 0: in-memory, size 0 B
I180720 07:16:36.717154 123057 server/node.go:421  [n?] **** cluster b2651e19-6f44-4455-9b69-55074045bd14 has been created
I180720 07:16:36.717395 123057 server/server.go:1370  [n?] **** add additional nodes by specifying --join=127.0.0.1:43885
I180720 07:16:36.739675 123057 storage/store.go:1480  [n1,s1] [n1,s1]: failed initial metrics computation: [n1,s1]: system config not yet available
I180720 07:16:36.743459 123057 server/node.go:553  [n1] initialized store [n1,s1]: disk (capacity=512 MiB, available=512 MiB, used=0 B, logicalBytes=6.9 KiB), ranges=1, leases=0, writes=0.00, bytesPerReplica={p10=7043.00 p25=7043.00 p50=7043.00 p75=7043.00 p90=7043.00 pMax=7043.00}, writesPerReplica={p10=0.00 p25=0.00 p50=0.00 p75=0.00 p90=0.00 pMax=0.00}
I180720 07:16:36.750288 123057 server/node.go:399  [n1] node ID 1 initialized
I180720 07:16:36.750917 123057 gossip/gossip.go:363  [n1] NodeDescriptor set to node_id:1 address:<network_field:"tcp" address_field:"127.0.0.1:43885" > attrs:<> locality:<> ServerVersion:<major_val:2 minor_val:0 patch:0 unstable:9 > 
I180720 07:16:36.758994 123057 storage/stores.go:222  [n1] read 0 node addresses from persistent storage
I180720 07:16:36.762549 123057 server/node.go:703  [n1] connecting to gossip network to verify cluster ID...
I180720 07:16:36.762856 123057 server/node.go:728  [n1] node connected via gossip and verified as part of cluster "b2651e19-6f44-4455-9b69-55074045bd14"
I180720 07:16:36.765648 123057 server/node.go:487  [n1] node=1: started with [<no-attributes>=<in-mem>] engine(s) and attributes []
I180720 07:16:36.782877 123057 server/config.go:336  [n1] available memory from cgroups (8.0 EiB) exceeds system memory 16 GiB, using system memory
I180720 07:16:36.784898 123057 server/server.go:1799  [n1] Could not start heap profiler worker due to: directory to store profiles could not be determined
I180720 07:16:36.790106 123057 server/server.go:1507  [n1] starting https server at 127.0.0.1:43733
I180720 07:16:36.790306 123057 server/server.go:1508  [n1] starting grpc/postgres server at 127.0.0.1:43885
I180720 07:16:36.790368 123057 server/server.go:1509  [n1] advertising CockroachDB node at 127.0.0.1:43885
W180720 07:16:36.790766 123057 sql/jobs/registry.go:299  [n1] unable to get node liveness: node not in the liveness table
I180720 07:16:37.212314 123428 rpc/nodedialer/nodedialer.go:90  [replica consistency checker,n1,s1,r1/1:/M{in-ax}] connection to n1 established
I180720 07:16:37.439144 123001 sql/event_log.go:126  [n1,intExec=optInToDiagnosticsStatReporting] Event: "set_cluster_setting", target: 0, info: {SettingName:diagnostics.reporting.enabled Value:true User:root}
I180720 07:16:37.905699 123446 sql/event_log.go:126  [n1,intExec=set-setting] Event: "set_cluster_setting", target: 0, info: {SettingName:version Value:$1 User:root}
I180720 07:16:38.177206 123005 sql/event_log.go:126  [n1,intExec=disableNetTrace] Event: "set_cluster_setting", target: 0, info: {SettingName:trace.debug.enable Value:false User:root}
I180720 07:16:38.574278 123476 sql/event_log.go:126  [n1,intExec=initializeClusterSecret] Event: "set_cluster_setting", target: 0, info: {SettingName:cluster.secret Value:gen_random_uuid()::STRING User:root}
I180720 07:16:38.795452 123410 sql/event_log.go:126  [n1,intExec=create-default-db] Event: "create_database", target: 50, info: {DatabaseName:defaultdb Statement:CREATE DATABASE IF NOT EXISTS defaultdb User:root}
I180720 07:16:38.946754 123462 sql/event_log.go:126  [n1,intExec=create-default-db] Event: "create_database", target: 51, info: {DatabaseName:postgres Statement:CREATE DATABASE IF NOT EXISTS postgres User:root}
I180720 07:16:39.021575 123057 server/server.go:1588  [n1] done ensuring all necessary migrations have run
I180720 07:16:39.021897 123057 server/server.go:1591  [n1] serving sql connections
I180720 07:16:39.066368 123057 storage/node_liveness_test.go:798  [in test] setting zone config to disable replication
I180720 07:16:39.066710 123057 storage/node_liveness_test.go:808  [in test] starting 3 more nodes
I180720 07:16:39.079803 123507 server/server_update.go:65  [n1] no need to upgrade, cluster already at the newest version
I180720 07:16:39.196410 123509 sql/event_log.go:126  [n1] Event: "node_join", target: 1, info: {Descriptor:{NodeID:1 Address:{NetworkField:tcp AddressField:127.0.0.1:43885} Attrs: Locality: ServerVersion:2.0-9} ClusterID:b2651e19-6f44-4455-9b69-55074045bd14 StartedAt:1532070996765127241 LastUp:1532070996765127241}
W180720 07:16:39.253982 123057 server/status/runtime.go:143  Could not parse build timestamp: parsing time "" as "2006/01/02 15:04:05": cannot parse "" as "2006"
I180720 07:16:39.428905 123057 server/server.go:795  [n?] monitoring forward clock jumps based on server.clock.forward_jump_check_enabled
I180720 07:16:39.546519 123057 server/config.go:552  [n?] 1 storage engine initialized
I180720 07:16:39.548580 123057 server/config.go:555  [n?] RocksDB cache size: 128 MiB
I180720 07:16:39.548662 123057 server/config.go:555  [n?] store 0: in-memory, size 0 B
W180720 07:16:39.549267 123057 gossip/gossip.go:1325  [n?] no incoming or outgoing connections
I180720 07:16:39.573745 123057 server/server.go:1372  [n?] no stores bootstrapped and --join flag specified, awaiting init command.
I180720 07:16:39.831393 123557 gossip/client.go:129  [n?] started gossip client to 127.0.0.1:43885
I180720 07:16:39.834889 123484 gossip/server.go:219  [n1] received initial cluster-verification connection from {tcp 127.0.0.1:41489}
I180720 07:16:39.851030 123057 storage/stores.go:222  [n?] read 0 node addresses from persistent storage
I180720 07:16:39.851468 123057 storage/stores.go:241  [n?] wrote 1 node addresses to persistent storage
I180720 07:16:39.851518 123057 server/node.go:703  [n?] connecting to gossip network to verify cluster ID...
I180720 07:16:39.851719 123057 server/node.go:728  [n?] node connected via gossip and verified as part of cluster "b2651e19-6f44-4455-9b69-55074045bd14"
I180720 07:16:39.855993 123498 kv/dist_sender.go:342  [n?] unable to determine this node's attributes for replica selection; node is most likely bootstrapping
I180720 07:16:39.891778 123497 kv/dist_sender.go:342  [n?] unable to determine this node's attributes for replica selection; node is most likely bootstrapping
I180720 07:16:39.906950 123057 kv/dist_sender.go:342  [n?] unable to determine this node's attributes for replica selection; node is most likely bootstrapping
I180720 07:16:39.921111 123057 server/node.go:392  [n?] new node allocated ID 2
I180720 07:16:39.922078 123057 gossip/gossip.go:363  [n2] NodeDescriptor set to node_id:2 address:<network_field:"tcp" address_field:"127.0.0.1:41489" > attrs:<> locality:<> ServerVersion:<major_val:2 minor_val:0 patch:0 unstable:9 > 
I180720 07:16:39.923280 123057 server/node.go:468  [n2] node=2: asynchronously bootstrapping engine(s) [<no-attributes>=<in-mem>]
I180720 07:16:39.925873 123057 server/node.go:487  [n2] node=2: started with [] engine(s) and attributes []
I180720 07:16:39.929280 123590 storage/stores.go:241  [n1] wrote 1 node addresses to persistent storage
I180720 07:16:39.930678 123057 server/config.go:336  [n2] available memory from cgroups (8.0 EiB) exceeds system memory 16 GiB, using system memory
I180720 07:16:39.931064 123057 server/server.go:1799  [n2] Could not start heap profiler worker due to: directory to store profiles could not be determined
I180720 07:16:39.948979 123057 server/server.go:1507  [n2] starting https server at 127.0.0.1:37401
I180720 07:16:39.950345 123057 server/server.go:1508  [n2] starting grpc/postgres server at 127.0.0.1:41489
I180720 07:16:39.950451 123057 server/server.go:1509  [n2] advertising CockroachDB node at 127.0.0.1:41489
I180720 07:16:39.999836 123057 server/server.go:1588  [n2] done ensuring all necessary migrations have run
I180720 07:16:40.000118 123057 server/server.go:1591  [n2] serving sql connections
W180720 07:16:40.151307 123057 server/status/runtime.go:143  Could not parse build timestamp: parsing time "" as "2006/01/02 15:04:05": cannot parse "" as "2006"
I180720 07:16:40.301966 123501 server/node.go:681  [n2] bootstrapped store [n2,s2]
I180720 07:16:40.364591 123591 server/server_update.go:65  [n2] no need to upgrade, cluster already at the newest version
I180720 07:16:40.413424 123057 server/server.go:795  [n?] monitoring forward clock jumps based on server.clock.forward_jump_check_enabled
I180720 07:16:40.446982 123057 server/config.go:552  [n?] 1 storage engine initialized
I180720 07:16:40.447193 123057 server/config.go:555  [n?] RocksDB cache size: 128 MiB
I180720 07:16:40.447238 123057 server/config.go:555  [n?] store 0: in-memory, size 0 B
W180720 07:16:40.447625 123057 gossip/gossip.go:1325  [n?] no incoming or outgoing connections
I180720 07:16:40.448427 123057 server/server.go:1372  [n?] no stores bootstrapped and --join flag specified, awaiting init command.
I180720 07:16:40.570511 123593 sql/event_log.go:126  [n2] Event: "node_join", target: 2, info: {Descriptor:{NodeID:2 Address:{NetworkField:tcp AddressField:127.0.0.1:41489} Attrs: Locality: ServerVersion:2.0-9} ClusterID:b2651e19-6f44-4455-9b69-55074045bd14 StartedAt:1532070999923781591 LastUp:1532070999923781591}
I180720 07:16:40.876232 123782 gossip/client.go:129  [n?] started gossip client to 127.0.0.1:43885
I180720 07:16:40.879570 123693 gossip/server.go:219  [n1] received initial cluster-verification connection from {tcp 127.0.0.1:32969}
I180720 07:16:40.899999 123057 storage/stores.go:222  [n?] read 0 node addresses from persistent storage
I180720 07:16:40.900541 123057 storage/stores.go:241  [n?] wrote 1 node addresses to persistent storage
I180720 07:16:40.900689 123057 server/node.go:703  [n?] connecting to gossip network to verify cluster ID...
I180720 07:16:40.900921 123057 server/node.go:728  [n?] node connected via gossip and verified as part of cluster "b2651e19-6f44-4455-9b69-55074045bd14"
I180720 07:16:40.904698 123697 kv/dist_sender.go:342  [n?] unable to determine this node's attributes for replica selection; node is most likely bootstrapping
I180720 07:16:40.912645 123696 kv/dist_sender.go:342  [n?] unable to determine this node's attributes for replica selection; node is most likely bootstrapping
I180720 07:16:40.921425 123769 storage/stores.go:241  [n?] wrote 2 node addresses to persistent storage
I180720 07:16:40.922264 123057 kv/dist_sender.go:342  [n?] unable to determine this node's attributes for replica selection; node is most likely bootstrapping
I180720 07:16:40.951644 123057 server/node.go:392  [n?] new node allocated ID 3
I180720 07:16:40.952447 123057 gossip/gossip.go:363  [n3] NodeDescriptor set to node_id:3 address:<network_field:"tcp" address_field:"127.0.0.1:32969" > attrs:<> locality:<> ServerVersion:<major_val:2 minor_val:0 patch:0 unstable:9 > 
I180720 07:16:40.953518 123057 server/node.go:468  [n3] node=3: asynchronously bootstrapping engine(s) [<no-attributes>=<in-mem>]
I180720 07:16:40.959258 123745 storage/stores.go:241  [n1] wrote 2 node addresses to persistent storage
I180720 07:16:40.963071 123746 storage/stores.go:241  [n2] wrote 2 node addresses to persistent storage
I180720 07:16:40.972791 123057 server/node.go:487  [n3] node=3: started with [] engine(s) and attributes []
I180720 07:16:40.980325 123057 server/config.go:336  [n3] available memory from cgroups (8.0 EiB) exceeds system memory 16 GiB, using system memory
I180720 07:16:40.981266 123057 server/server.go:1799  [n3] Could not start heap profiler worker due to: directory to store profiles could not be determined
I180720 07:16:41.016798 123057 server/server.go:1507  [n3] starting https server at 127.0.0.1:40475
I180720 07:16:41.016967 123057 server/server.go:1508  [n3] starting grpc/postgres server at 127.0.0.1:32969
I180720 07:16:41.017044 123057 server/server.go:1509  [n3] advertising CockroachDB node at 127.0.0.1:32969
I180720 07:16:41.072439 123057 server/server.go:1588  [n3] done ensuring all necessary migrations have run
I180720 07:16:41.072681 123057 server/server.go:1591  [n3] serving sql connections
I180720 07:16:41.417097 123801 server/node.go:681  [n3] bootstrapped store [n3,s3]
W180720 07:16:41.461153 123057 server/status/runtime.go:143  Could not parse build timestamp: parsing time "" as "2006/01/02 15:04:05": cannot parse "" as "2006"
I180720 07:16:41.578838 123876 server/server_update.go:65  [n3] no need to upgrade, cluster already at the newest version
I180720 07:16:41.641698 123878 sql/event_log.go:126  [n3] Event: "node_join", target: 3, info: {Descriptor:{NodeID:3 Address:{NetworkField:tcp AddressField:127.0.0.1:32969} Attrs: Locality: ServerVersion:2.0-9} ClusterID:b2651e19-6f44-4455-9b69-55074045bd14 StartedAt:1532071000953881669 LastUp:1532071000953881669}
I180720 07:16:41.711247 123057 server/server.go:795  [n?] monitoring forward clock jumps based on server.clock.forward_jump_check_enabled
I180720 07:16:41.779071 123057 server/config.go:552  [n?] 1 storage engine initialized
I180720 07:16:41.779889 123057 server/config.go:555  [n?] RocksDB cache size: 128 MiB
I180720 07:16:41.779961 123057 server/config.go:555  [n?] store 0: in-memory, size 0 B
W180720 07:16:41.809010 123057 gossip/gossip.go:1325  [n?] no incoming or outgoing connections
I180720 07:16:41.811024 123057 server/server.go:1372  [n?] no stores bootstrapped and --join flag specified, awaiting init command.
I180720 07:16:42.050146 123561 gossip/client.go:129  [n?] started gossip client to 127.0.0.1:43885
I180720 07:16:42.059708 124056 gossip/server.go:219  [n1] received initial cluster-verification connection from {tcp 127.0.0.1:38351}
I180720 07:16:42.120383 123057 storage/stores.go:222  [n?] read 0 node addresses from persistent storage
I180720 07:16:42.122169 123057 storage/stores.go:241  [n?] wrote 3 node addresses to persistent storage
I180720 07:16:42.122247 123057 server/node.go:703  [n?] connecting to gossip network to verify cluster ID...
I180720 07:16:42.122527 123057 server/node.go:728  [n?] node connected via gossip and verified as part of cluster "b2651e19-6f44-4455-9b69-55074045bd14"
I180720 07:16:42.133967 123970 kv/dist_sender.go:342  [n?] unable to determine this node's attributes for replica selection; node is most likely bootstrapping
I180720 07:16:42.142438 123969 kv/dist_sender.go:342  [n?] unable to determine this node's attributes for replica selection; node is most likely bootstrapping
I180720 07:16:42.161098 123057 kv/dist_sender.go:342  [n?] unable to determine this node's attributes for replica selection; node is most likely bootstrapping
I180720 07:16:42.174335 123057 server/node.go:392  [n?] new node allocated ID 4
I180720 07:16:42.175091 123057 gossip/gossip.go:363  [n4] NodeDescriptor set to node_id:4 address:<network_field:"tcp" address_field:"127.0.0.1:38351" > attrs:<> locality:<> ServerVersion:<major_val:2 minor_val:0 patch:0 unstable:9 > 
I180720 07:16:42.176193 123057 server/node.go:468  [n4] node=4: asynchronously bootstrapping engine(s) [<no-attributes>=<in-mem>]
I180720 07:16:42.176888 123057 server/node.go:487  [n4] node=4: started with [] engine(s) and attributes []
I180720 07:16:42.179294 123057 server/config.go:336  [n4] available memory from cgroups (8.0 EiB) exceeds system memory 16 GiB, using system memory
I180720 07:16:42.180732 123057 server/server.go:1799  [n4] Could not start heap profiler worker due to: directory to store profiles could not be determined
I180720 07:16:42.196507 124070 storage/stores.go:241  [n1] wrote 3 node addresses to persistent storage
I180720 07:16:42.203667 124023 storage/stores.go:241  [n2] wrote 3 node addresses to persistent storage
I180720 07:16:42.205206 124058 storage/stores.go:241  [n3] wrote 3 node addresses to persistent storage
I180720 07:16:42.247602 123057 server/server.go:1507  [n4] starting https server at 127.0.0.1:38091
I180720 07:16:42.247823 123057 server/server.go:1508  [n4] starting grpc/postgres server at 127.0.0.1:38351
I180720 07:16:42.247878 123057 server/server.go:1509  [n4] advertising CockroachDB node at 127.0.0.1:38351
W180720 07:16:42.252559 123057 sql/jobs/registry.go:299  [n4] unable to get node liveness: node not in the liveness table
I180720 07:16:42.285576 123057 server/server.go:1588  [n4] done ensuring all necessary migrations have run
I180720 07:16:42.285802 123057 server/server.go:1591  [n4] serving sql connections
I180720 07:16:42.301531 123057 storage/node_liveness_test.go:813  [in test] waiting for node statuses
I180720 07:16:42.349863 124094 server/node.go:681  [n4] bootstrapped store [n4,s4]
I180720 07:16:42.589042 124212 server/server_update.go:65  [n4] no need to upgrade, cluster already at the newest version
I180720 07:16:42.623966 124214 sql/event_log.go:126  [n4] Event: "node_join", target: 4, info: {Descriptor:{NodeID:4 Address:{NetworkField:tcp AddressField:127.0.0.1:38351} Attrs: Locality: ServerVersion:2.0-9} ClusterID:b2651e19-6f44-4455-9b69-55074045bd14 StartedAt:1532071002176552750 LastUp:1532071002176552750}
I180720 07:16:43.143003 123057 storage/node_liveness_test.go:815  [in test] waiting done
I180720 07:16:43.143457 123057 storage/node_liveness_test.go:832  [in test] shutting down node 2
I180720 07:16:43.162023 123057 storage/node_liveness_test.go:834  [in test] done shutting down node 2
I180720 07:16:43.162354 123057 storage/node_liveness_test.go:837  [in test] decommissioning node 3
I180720 07:16:43.232403 123057 sql/event_log.go:126  [in test] Event: "node_decommissioned", target: 3, info: {}
I180720 07:16:43.233085 123057 storage/node_liveness_test.go:841  [in test] done decommissioning node 3
I180720 07:16:43.233408 123057 storage/node_liveness_test.go:844  [in test] decommissioning and shutting down node 4
I180720 07:16:43.347078 123057 sql/event_log.go:126  [in test] Event: "node_decommissioned", target: 4, info: {}
I180720 07:16:43.374796 123057 storage/node_liveness_test.go:849  [in test] done removing node 4
I180720 07:16:43.375062 123057 storage/node_liveness_test.go:851  [in test] checking status map

```

The test also has an unrelated flake:

```
make stress TESTTIMEOUT=1m PKG=./pkg/storage/ TESTS=TestNodeLivenessStatusMap STRESSFLAGS='-timeout 30s -kill'
--- FAIL: TestNodeLivenessStatusMap (0.52s)
	node_liveness_test.go:846: during liveness update 4 -> true: node not in the liveness table
```

--- a/pkg/server/server.go
+++ b/pkg/server/server.go
@@ -1541,9 +1541,12 @@ If problems persist, please see ` + base.DocsURL("cluster-setup-troubleshooting.
 			select {
 			case decommissionSem <- struct{}{}:
 				s.stopper.RunWorker(ctx, func(context.Context) {
-					// Don't use the passed in ctx because there is an associated timeout
-					// meant to be used when heartbeating.
-					ctx := context.Background()
+					// Don't use the passed in ctx because there is an associated
+					// timeout meant to be used when heartbeating. Instead, tie the
+					// context to the stopper so that the drain attempt stops without
+					// blocking if the server starts shutting down.
+					ctx, done := s.stopper.WithCancelOnQuiesce(context.Background())
+					defer done()
 
 					defer func() {
 						<-decommissionSem

